### PR TITLE
Fix stale tiles being shown when calling VectorTileSource#setTiles while the map is moving

### PR DESCRIPTION
**Pull request by vanilla-lake**
<changelog>Fix stale tiles being shown when calling VectorTileSource#setTiles while the map is moving</changelog>

Before this PR, when calling `VectorTileSource#setTiles`, the source cache was cleared before the TileJSON was reloaded asynchronously. If the map was moving while `VectorTileSource#setTiles` was called, it was possible for the cache to be repopulated with the old tiles if `VectorTileSource#loadTile` happened to be called before the TileJSON had been reloaded. This was due to the `VectorTileSource#tiles` property, which is used by `VectorTileSource#loadTile`, only being updated asynchronously after the TileJSON had reloaded.

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Write tests for all new functionality.
 - [x] Manually test the debug page.
 - [x] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

### CODE REVIEW

File: src/source/vector_tile_source.ts [hunk 1]
```diff
@@ -100,6 +100,7 @@ class VectorTileSource extends Evented implements Source {
             if (err) {
                 this.fire(new ErrorEvent(err));
             } else if (tileJSON) {
+                this.map.style.sourceCaches[this.id].clearTiles();
```
Comment (HarelM) [inline src/source/vector_tile_source.ts:103]: It seems that before this change the cache will be cleared even if there was an error.
Is this intended in this change?

**Comment by HarelM** [comment]
Besides the comment above I think this looks good.
I think that someone else should go over it besides me though... @birkskyum @wipfli

File: src/source/vector_tile_source.ts (hunk 1 above)
Comment (vanilla-lake) [reply src/source/vector_tile_source.ts:103]: You're right... Thanks for pointing that out. I'll move this statement outside of the `if`.

**Review by birkskyum — APPROVED** [review]
Thanks! With the latest commit it lgtm.

**Comment by vanilla-lake** [comment]
No problem! 😄

**Comment by HarelM** [comment]
@vanilla-lake I'm running the benchmark and i see the following error in the console:
image
I have a feeling this is related to this change.
Can you please check if this is caused by this PR by any chance?
To see this do: `npm run start-bench` and open the browser at `http://localhost:9966/bench/versions` and wait for a while.

**Comment by wipfli** [comment]
I think I saw a similar error message with the headless version `npm run benchmarks`. Also there, you have to wait a while until this happens.

**Comment by vanilla-lake** [comment]
The issue is indeed related to the changes in this PR. Sorry about that... I created PR #984 which should hopefully resolve the issue.

**Comment by khmm12** [comment]
@vanilla-lake the issue is also reproducible in "real world"

**Comment by vanilla-lake** [comment]
Thanks for the fix @khmm12! Your fix seems correct, since calling `SourceCache.onRemove()` will call `VectorTileSource.onRemove()` which will itself cancel `_tileJSONRequest`.

This error could happen anytime `Map.style` becomes `null` or `undefined`. As far as I can see, the only ways this can happen is if `Map.setStyle()` is called with a `null` or `undefined` argument. In this case, `Style._remove()` should be called on the existing style from `Map._updateStyle()`, now cancelling any pending TileJSON requests thanks to your fix.

**Pull request merged**